### PR TITLE
Window based compaction

### PR DESCRIPTION
**Pull request by adam-yorwerth**
Compacts all messages prior to the configurable window, run via a configurable task.

### CODE REVIEW

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 1]
```diff
@@ -184,6 +181,17 @@ private PreparedStatement getMessagesStatement(Connection connection, Map<String
         }
     }
 
+    public void compactUpTo(ZonedDateTime threasholdDate) {
```
Comment (cwebbtw) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java]: threasholdDate typo

**Comment by KrzysztofKowalczyk** [comment]
I would not call it a window, window usually have a beginning and end

File: pipe-http-server-cloud/configuration/application.yml [hunk 2]
```diff
@@ -18,6 +18,9 @@ pipe:
     url: "{PIPE_URL}"
 
 persistence:
+  compact:
+    threshold: 3d
+    schedule.cron: "0 14 3 1/1 * ?"
```
Comment (KrzysztofKowalczyk) [inline pipe-http-server-cloud/configuration/application.yml]: I would rather prefer to run at different hour than tills, also tills for supportability sake.
In long run I would like tills to spread their runs not to run at the same time everywhere.

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 3]
```diff
@@ -216,4 +224,14 @@ private static String getSelectLatestOffsetWithoutTypeQuery(boolean tags) {
...
     }
+
+    private static String getCompactionQuery() {
+        return "DELETE FROM events " +
+                "where msg_offset in " +
+                "(select msg_offset from events e, " +
+                "( SELECT msg_key, max(msg_offset) " +
+                "max_offset FROM events GROUP BY msg_key) " +
+                "x where created_utc <= ? " +
+                "and e.msg_key = x.msg_key and e.msg_offset <> x.max_offset);";
```
Comment (KrzysztofKowalczyk) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java]: I might prefer the slower query as it it easier to understand

File: pipe-http-server-cloud/configuration/application.yml (hunk 2 above)
Comment (adam-yorwerth) [reply pipe-http-server-cloud/configuration/application.yml]: Changed to 2:14am

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java (hunk 3 above)
Comment (adam-yorwerth) [reply pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java]: I think performance is the priority here

**Pull request merged**